### PR TITLE
Add `HighestResponseRatioNextScheduling.java` new algorithm with tests

### PR DESCRIPTION
**Pull request by Hardvan**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+ * This algorithm aims to reduce the average waiting time
+ * and improve overall system performance by balancing short and long processes.
+ */
+public final class HighestResponseRatioNextScheduling {
+    private HighestResponseRatioNextScheduling() {
+    }
+
+    /**
+     * Calculates the Turn Around Time (TAT) for each process.
+     *
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:21]: Good to highlight how Turn Around Time is calculated and what is this, with some formatting.

Like a variant like this, but maybe more detailed, for better understanding:
```
     * <p>Turn Around Time is calculated as the total time a process spends 
     * in the system from arrival to completion.</p>
```

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+     * @param currentTime     The current time in the scheduling process.
+     * @return The index of the process with the highest response ratio.
+     */
+    private static int findHighestResponseRatio(String[] processNames, int[] arrivalTimes, int[] burstTimes, boolean[] finishedProcess, int currentTime) {
+        double maxResponseRatio = -1;
+        int loc = -1;
+
+        for (int i = 0; i < processNames.length; i++) {
+            if (!finishedProcess[i] && arrivalTimes[i] <= currentTime) {
+                double responseRatio = (double) (burstTimes[i] + currentTime - arrivalTimes[i]) / burstTimes[i];
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:116]: Need to mention this in method description

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+     *
+     * @param processNames    Array of process names.
+     * @param arrivalTimes    Array of arrival times for each process.
+     * @param burstTimes      Array of burst times for each process.
+     * @param finishedProcess Array indicating whether each process has finished.
+     * @param currentTime     The current time in the scheduling process.
+     * @return The index of the process with the highest response ratio.
+     */
+    private static int findHighestResponseRatio(String[] processNames, int[] arrivalTimes, int[] burstTimes, boolean[] finishedProcess, int currentTime) {
+        double maxResponseRatio = -1;
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:111]: What do you think about moving all these default values ​​into constants with properly described names for better understanding?

File: src/test/java/com/thealgorithms/scheduling/HighestResponseRatioNextSchedulingTest.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import org.junit.jupiter.api.Test;
+
+public class HighestResponseRatioNextSchedulingTest {
+
+    @Test
+    public void testCalculateTurnAroundTime() {
+        String[] processNames = {"A", "B", "C"};
+        int[] arrivalTimes = {0, 2, 4};
+        int[] burstTimes = {3, 1, 2};
+        int noOfProcesses = 3;
```
Comment (alxkm) [inline src/test/java/com/thealgorithms/scheduling/HighestResponseRatioNextSchedulingTest.java:14]: It's good to have a test for when this value is zero, and other tests that will cover edge cases.

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+     * @param currentTime     The current time in the scheduling process.
+     * @return The index of the next process to be scheduled.
+     */
+    private static int findNextProcess(int[] arrivalTimes, boolean[] finishedProcess, int currentTime) {
+        for (int i = 0; i < arrivalTimes.length; i++) {
+            if (!finishedProcess[i] && arrivalTimes[i] <= currentTime) {
+                return i;
+            }
+        }
+        return 0;
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:97]: Can this value, cause problems if the process with index 0 is not ready yet?

If yes, for such case better instead current approach, to return -1 or another special index to explicitly indicate that the process was not found. In the loop body, it is worth checking that the process actually exists.

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    /**
+     * Calculates the Turn Around Time (TAT) for each process.
+     *
+     * @param processNames Array of process names.
+     * @param arrivalTimes Array of arrival times corresponding to each process.
+     * @param burstTimes   Array of burst times for each process.
+     * @param noOfProcesses The number of processes.
+     * @return An array of Turn Around Times for each process.
+     */
+    public static int[] calculateTurnAroundTime(String[] processNames, int[] arrivalTimes, int[] burstTimes, int noOfProcesses) {
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:28]: Some of these parameters in another method may be final to emphasize that they are constant within the method.

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    /**
+     * Sorts the indices of the arrival times array in ascending order.
+     *
+     * @param arrivalTimes Array of arrival times for each process.
+     * @return An array of indices sorted by the corresponding arrival times.
+     */
+    private static int[] sortIndicesByArrival(int[] arrivalTimes) {
+        Integer[] indices = new Integer[arrivalTimes.length];
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:133]: Better to use `int[]` instead of `Integer[]` to avoid  unnecessary boxing/unboxing. Since method return `int[]`.

Or maybe it can be better rewritten to kind of such implementation:

```
private static int[] sortIndicesByArrival(int[] arrivalTimes) {
    return IntStream.range(0, arrivalTimes.length)
        .boxed()
        .sorted(Comparator.comparingInt(i -> arrivalTimes[i]))
        .mapToInt(Integer::intValue)
        .toArray();
}
```

What do you think?

**Review by alxkm — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+     *
+     * @param processNames Array of process names.
+     * @param arrivalTimes Array of arrival times corresponding to each process.
+     * @param burstTimes   Array of burst times for each process.
+     * @param noOfProcesses The number of processes.
+     * @return An array of Turn Around Times for each process.
+     */
+    public static int[] calculateTurnAroundTime(String[] processNames, int[] arrivalTimes, int[] burstTimes, int noOfProcesses) {
+        int currentTime = 0;
+        int[] turnAroundTime = new int[noOfProcesses];
```
[the remark this answers is not in the document]
Comment (Hardvan) [reply src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:30]: `Process` class implemented.

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * The HighestResponseRatioNextScheduling class implements
+ * the Highest Response Ratio Next (HRRN) scheduling algorithm.
+ * HRRN is a non-preemptive scheduling algorithm that
+ * selects the process with the highest response ratio for execution.
+ * The response ratio is calculated as (waiting time + burst time) / burst time.
+ * This algorithm aims to reduce the average waiting time
+ * and improve overall system performance by balancing short and long processes.
+ */
```
[the remark this answers is not in the document]
Comment (Hardvan) [reply src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:14]: Done!

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java (hunk 1 above)
Comment (Hardvan) [reply src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:21]: Done!

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java (hunk 2 above)
Comment (Hardvan) [reply src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:116]: Done!

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java (hunk 3 above)
Comment (Hardvan) [reply src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:111]: That would indeed be more readable and efficient. I have declared two static constants:

```java
private static final int PROCESS_NOT_FOUND = -1;
private static final double INITIAL_MAX_RESPONSE_RATIO = -1.0;
```

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java (hunk 5 above)
Comment (Hardvan) [reply src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:97]: I have changed this to the following:

```java
int maxIndex = PROCESS_NOT_FOUND; // PROCESS_NOT_FOUND = -1
...
// Find the process with the highest response ratio
...
return maxIndex;
```

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java (hunk 6 above)
Comment (Hardvan) [reply src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:28]: Noted, I have changed the function header to the following:

```java
public static int[] calculateTurnAroundTime(final String[] processNames, final int[] arrivalTimes, final int[] burstTimes, final int noOfProcesses)
```

File: src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java (hunk 7 above)
Comment (Hardvan) [reply src/main/java/com/thealgorithms/scheduling/HighestResponseRatioNextScheduling.java:133]: Insights like these is why I like contributing to open source and learning valuable information!
I simplified the function call to the following:

```java
 public static int[] calculateTurnAroundTime(final String[] processNames, final int[] arrivalTimes, final int[] burstTimes, final int noOfProcesses) {
	...
	Arrays.sort(processes, Comparator.comparingInt(p -> p.arrivalTime));
	...;
```

File: src/test/java/com/thealgorithms/scheduling/HighestResponseRatioNextSchedulingTest.java (hunk 4 above)
Comment (Hardvan) [reply src/test/java/com/thealgorithms/scheduling/HighestResponseRatioNextSchedulingTest.java:14]: Done! I added the following additional test cases:

```java
@Test
    public void testZeroProcesses() {
        String[] processNames = {};
        int[] arrivalTimes = {};
        int[] burstTimes = {};
        int noOfProcesses = 0;

        int[] expectedTurnAroundTimes = {};
        int[] actualTurnAroundTimes = HighestResponseRatioNextScheduling.calculateTurnAroundTime(processNames, arrivalTimes, burstTimes, noOfProcesses);

        assertArrayEquals(expectedTurnAroundTimes, actualTurnAroundTimes, "Turn Around Times for zero processes should be an empty array");
    }

    @Test
    public void testAllProcessesArriveAtSameTime() {
        String[] processNames = {"A", "B", "C", "D"};
        int[] arrivalTimes = {0, 0, 0, 0};
        int[] burstTimes = {4, 3, 1, 2};
        int noOfProcesses = 4;

        int[] expectedTurnAroundTimes = {4, 10, 5, 7};
        int[] actualTurnAroundTimes = HighestResponseRatioNextScheduling.calculateTurnAroundTime(processNames, arrivalTimes, burstTimes, noOfProcesses);

        assertArrayEquals(expectedTurnAroundTimes, actualTurnAroundTimes, "Turn Around Times for processes arriving at the same time do not match");
    }

    @Test
    public void testProcessesWithZeroBurstTime() {
        String[] processNames = {"A", "B", "C"};
        int[] arrivalTimes = {0, 1, 2};
        int[] burstTimes = {3, 0, 2};
        int noOfProcesses = 3;

        int[] expectedTurnAroundTimes = {3, 2, 3};
        int[] actualTurnAroundTimes = HighestResponseRatioNextScheduling.calculateTurnAroundTime(processNames, arrivalTimes, burstTimes, noOfProcesses);

        assertArrayEquals(expectedTurnAroundTimes, actualTurnAroundTimes, "Turn Around Times for processes with zero burst time do not match");
    }

    @Test
    public void testProcessesWithLargeGapsBetweenArrivals() {
        String[] processNames = {"A", "B", "C"};
        int[] arrivalTimes = {0, 100, 200};
        int[] burstTimes = {10, 10, 10};
        int noOfProcesses = 3;

        int[] expectedTurnAroundTimes = {10, 10, 10};
        int[] actualTurnAroundTimes = HighestResponseRatioNextScheduling.calculateTurnAroundTime(processNames, arrivalTimes, burstTimes, noOfProcesses);

        assertArrayEquals(expectedTurnAroundTimes, actualTurnAroundTimes, "Turn Around Times for processes with large gaps between arrivals do not match");
    }

    @Test
    public void testProcessesWithVeryLargeBurstTimes() {
        String[] processNames = {"A", "B"};
        int[] arrivalTimes = {0, 1};
        int[] burstTimes = {Integer.MAX_VALUE/2, Integer.MAX_VALUE/2};
        int noOfProcesses = 2;

        int[] expectedTurnAroundTimes = {Integer.MAX_VALUE/2, Integer.MAX_VALUE-2};
        int[] actualTurnAroundTimes = HighestResponseRatioNextScheduling.calculateTurnAroundTime(processNames, arrivalTimes, burstTimes, noOfProcesses);

        assertArrayEquals(expectedTurnAroundTimes, actualTurnAroundTimes, "Turn Around Times for processes with very large burst times do not match");
    }
```

**Comment by Hardvan** [comment]
The suggested changes have been implemented.

**Review by alxkm — APPROVED** [review]
